core/services/pipeline/runner.go

```diff
@@ -199,25 +171,117 @@ func (r *runner) panickedRunResults(spec Spec) (Run, []TaskRunResult, error) {
 	run.PipelineSpecID = spec.ID
 	run.CreatedAt = time.Now()
 	run.FinishedAt = &run.CreatedAt
-	tasks, err := spec.TasksInDependencyOrder()
+	p, err := spec.Pipeline()
 	if err != nil {
 		return run, nil, err
 	}
 	f := time.Now()
-	for _, task := range tasks {
+	for _, task := range p.Tasks {
 		panickedTrrs = append(panickedTrrs, TaskRunResult{
 			Task:       task,
 			CreatedAt:  f,
 			Result:     Result{Value: nil, Error: ErrRunPanicked},
 			FinishedAt: time.Now(),
-			IsTerminal: task.OutputTask() == nil,
 		})
 	}
 	run.Outputs = TaskRunResults(panickedTrrs).FinalResult().OutputsDB()
 	run.Errors = TaskRunResults(panickedTrrs).FinalResult().ErrorsDB()
 	return run, panickedTrrs, nil
 }
 
+type scheduler struct {
+	pipeline     *Pipeline
+	dependencies map[int64]uint
+	input        interface{}
+	waiting      uint
+	// roots are the tasks at the start of the pipeline
+	roots   []Task
+	results map[int64]TaskRunResult
+
+	taskCh   chan *memoryTaskRun
+	resultCh chan TaskRunResult
+}
+
+func newScheduler(p *Pipeline, i interface{}) *scheduler {
+	dependencies := make(map[int64]uint, len(p.Tasks))
+	var roots []Task
+
+	for id, task := range p.Tasks {
+		i := len(task.Inputs())
+		dependencies[id] = uint(i)
+
+		// no inputs: this is a root
+		if i == 0 {
+			roots = append(roots, task)
+		}
+	}
+	s := &scheduler{
+		pipeline:     p,
+		dependencies: dependencies,
+		input:        i,
+		results:      make(map[int64]TaskRunResult, len(p.Tasks)),
+
+		// taskCh should never block
+		taskCh:   make(chan *memoryTaskRun, len(dependencies)),
+		resultCh: make(chan TaskRunResult),
+	}
+
+	for _, task := range roots {
+		run := &memoryTaskRun{task: task}
+		// fill in the inputs
+		run.inputs = append(run.inputs, input{index: 0, result: Result{Value: s.input}})
+
+		s.taskCh <- run
+		s.waiting++
+	}
+
+	go s.run()
```

Comment:
Is it confusing to have a `newXXX` function also spin off a goroutine? Means that instantiation is tied to booting which can be annoying to work with